shocker.c

```diff
@@ -1,60 +1,161 @@
 #define _GNU_SOURCE
-#include <sys/wait.h> // waitpid
+#include <stdlib.h>
+#include <stdbool.h>
 #include <sched.h> // clone, all the clone Flags
 #include <unistd.h> // execvp, sethostname
 #include <string.h> // strlen
 #include <stdio.h> // puts, printf
 #include <sys/mount.h> // mount
+#include <sys/stat.h> // mkdir
+#include <sys/types.h> // Flags for sys stuff
+#include <sys/mman.h>
+#include <sys/wait.h> // waitpid
 
-#define STACK_SIZE (1024 * 1024)
+#define STACK_SIZE (1024 * 1024)    /* Stack size for cloned child */
 
 //   docker run image <cmd> <params>
 //./shocker run       <cmd> <params>
 
-static char child_stack[STACK_SIZE]; /* Stack size for cloned child */
-
 int child(void *arg) {
-  sethostname("Container", 10);
-  chdir("/root/container");
-  chroot("/root/container"); // this is the main part of hiding
-  mount("proc", "/proc", "proc", 0, NULL); // proc is a pseudo-fs, used by many tools
+  if (sethostname("Container", 10) == -1) {
+    perror("sethostname failed with: ");
+    goto fail_sethostname;
+  }
+
+  if (chdir("/root/container") == -1) {
+    perror("sethostname failed with: ");
+    goto fail_chdir;
+  }
+
+  if (chroot("/root/container") == -1) {
+    perror("sethostname failed with: ");
+    goto fail_chroot;
+  }
+
+  if (mount("proc", "/proc", "proc", 0, NULL) == -1) {
+    perror("mount failed with: ");
+    goto fail_mount;
+  }
+
   char **shell = (char **)arg;
```

Comment:
Casting to and from void can be done implicitly. You don't need to make the cast explicit. this mostly boils down to a stylistic choice.

Reply:
I index into `shell` though, and that leads to an error if not cast.

Reply:
What I was suggesting is doing the following: 
```suggestion
  char **shell =arg;
```
The cast will be done implicitly to and from `void *` but you still need to set the right type.

Reply:
derp